File: src/app/useCases/trafficSource/updateTrafficSourceUseCase.ts
import { UserGatewayDTO } from "../../../domain/gateways/user";
import { TrafficSourceRepository } from "../../../domain/repositories/trafficSource";
import { HttpAdapter } from "../../../infra/adapters/httpAdapter";

type InputProps = {
  trafficSourceId: string;
  name?: string;
  trafficDomain?: string;
};

class UpdateTrafficSourceUseCase {
  constructor(
    private trafficSourceRepository: TrafficSourceRepository,
    private userGateway: UserGatewayDTO
  ) {}

  async execute(input: InputProps, token: string) {
    const { trafficSourceId, name, trafficDomain } = input;

    const [trafficSource, user] = await Promise.all([
      this.trafficSourceRepository.findById(trafficSourceId),
      this.userGateway.findUnique(token),
    ]);

    if (!trafficSource) {
      throw HttpAdapter.notFound("Traffic source not found");
    }

    if (trafficSource.userId !== user.id) {
      throw HttpAdapter.forbidden("You do not own this traffic source.");
    }

    trafficSource.update({ name, trafficDomain });

    await this.trafficSourceRepository.updateTrafficSource(trafficSource);

    return trafficSource.toJson(user.utc);
  }
}

export { UpdateTrafficSourceUseCase };

File: internal/app/useCases/trafficSource/updateTrafficSourceUseCase.go
package trafficSourceUseCases

import (
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/entities"
	domainErrors "github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/errors"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/providers"
	"github.com/Lucas-Eduardo-Goncalves/arkyn-store/internal/domain/repositories"
)

type UpdateTrafficSourceUseCase struct {
	trafficSourceRepository repositories.TrafficSourceRepository
	dateGenerator           providers.DateGenerator
}

type UpdateTrafficSourceInput struct {
	Name            *string
	TrafficDomain   *string
	UserId          string
	TrafficSourceId string
}

func NewUpdateTrafficSourceUseCase(
	trafficSourceRepository repositories.TrafficSourceRepository,
	dateGenerator providers.DateGenerator,
) *UpdateTrafficSourceUseCase {
	return &UpdateTrafficSourceUseCase{
		trafficSourceRepository: trafficSourceRepository,
		dateGenerator:           dateGenerator,
	}
}

func (t *UpdateTrafficSourceUseCase) Handle(input *UpdateTrafficSourceInput) (*entities.TrafficSource, error) {
	trafficSource, err := t.trafficSourceRepository.FindById(input.TrafficSourceId)

	if err != nil {
		return nil, err
	}

	if trafficSource == nil {
		return nil, domainErrors.NotFound("This traffic source don't exists.")
	}

	if trafficSource.UserId != input.UserId {
		return nil, domainErrors.Invalid("You aren't traffic source owner.")
	}

	trafficSource.Update(
		&entities.UpdateTrafficSourceInput{Name: input.Name, TrafficDomain: input.TrafficDomain},
		t.dateGenerator,
	)

	_, err = t.trafficSourceRepository.Update(trafficSource)

	if err != nil {
		return nil, err
	}

	return trafficSource, nil
}
